module/zfs/vdev_queue.c

```diff
@@ -535,16 +535,17 @@ vdev_queue_pending_remove(vdev_queue_t *vq, zio_t *zio)
 static void
 vdev_queue_agg_io_done(zio_t *aio)
 {
-	if (aio->io_type == ZIO_TYPE_READ) {
+	abd_put(aio->io_abd);
+	if (aio->io_type == ZIO_TYPE_WRITE) {
 		zio_t *pio;
 		zio_link_t *zl = NULL;
 		while ((pio = zio_walk_parents(aio, &zl)) != NULL) {
-			abd_copy_off(pio->io_abd, aio->io_abd,
-			    0, pio->io_offset - aio->io_offset, pio->io_size);
+			if (pio->io_flags & ZIO_FLAG_NODATA) {
+				abd_put(pio->io_abd);
```

Comment:
We got this abd from `abd_get_zeros()`, which gives us our own abd, which is why we need to abd_put() it.  Why can't we instead use `abd_add_child(manage_mem=B_TRUE)` to have the multi-abd infrastructure free it for us?

Reply:
I actually was looking at this earlier today and was thinking I can clean all this up. I need to clean up the code in vdev_queue_aggregate() for adding abd_get_zeros() to the multilist abd. Then in the vdev_queue_aggregate_done() function we just need to call abd_free().

Reply:
I wound up cleaning up the code in vdev_queue_aggregate() and now vdev_queue_agg_io_done() just consists of a call to abd_free().